### ISSUE TITLE
Consider encoding mesh consistency into types.

### ISSUE THREAD

**Issue by olson-sean-k** [olson-sean-k/plexus#12]
Perhaps mesh consistency should affect the mesh API. Today, mutations dance around operations that may or may not be safe due to the consistency of a mesh, such as implicitly or explicitly unwrapping connected topology or using circulators. Topological views provide a "raw" API that is available to the `graph` module, but code must "just know" when to use this instead of the user-facing API.

New types (maybe via additional type parameters) could differentiate consistent and (potentially) inconsistent meshes. Users would only ever see friendly consistent meshes and their related types. Internally, mutation code would be prevented from performing potentially unsound operations and, at the very least, would be forced to explicitly handle `Result`s and `Option`s.

Mutations would need to be based entirely on ownership or `mem::replace` (to support `&mut`), because a consistent mesh and an inconsistent mesh would be entirely different types and mutations would need to transition between them. It should be relatively inexpensive to move (shallow copy) a `Mesh`, because it consists of only three `HashMap`s. If it becomes too expensive, then that data can be boxed and placed into a container type.

Mutations would need to be seriously refactored if they are no longer able to freely traverse topology. Instead, any required keys or other information would need to be gathered before a mesh goes into a mutable, inconsistent state.

**Comment by olson-sean-k** [comment]
I've been experimenting quite a bit with this in a branch. It's messy, but I think encoding this information into types could be useful.

As of this writing, the `mutation` module could use some serious rethinking. [This gist](https://gist.github.com/olson-sean-k/7690ee7623aa10f4a94d14794ce0ed38) illustrates some ideas that could be useful. Here are a few important points:

- There is only one type that implements mutations: `Mutation`. There is no modality, and all mutations are batched. This avoids duplication and can be less performant, but it is important to generalize consistency checks across all types of topology.
- Mutations happen in three places. First, `Mutation` provides the most basic operations for insertion and removal and performs consistency checks when committed. Modules for each topology type (e.g., faces) provide higher order mutations as functions over `Mutation`s, allowing for batching. Finally, topological view types (i.e., `FaceMut`) expose those mutations to users.
- Cache types prepare topological data needed for high-order mutations. Mutation functions either operate on a cache or generate the cache on demand. This is a fragile API, as the cache must be derived from the correct `Mesh` instance and can become stale. However, this is an internal API, and allows for more complex batching and better performance when it is possible to cache data from a mesh while it is consistent and defer mutations.
- Topological view types expose a "reachable" API that is generic with respect to consistency. These are "best effort" traversals, and use fallback techniques that may be slower and are not necessary for consistent meshes (such as exploiting the relationship between vertex and edge keys). This is tricky, because it may not be clear if mutations performed against topology data gathered in this was is actually consistent. Whenever possible, consistent queries should be used before performing mutations. This is necessary to bridge the gap when a batch mutation needs additional topological data but cannot commit yet. This should of course be avoided as much as possible.
- Errors emitted when building a cache or performing a mutation via a module-level function should always be recoverable (i.e., nothing has actually mutated the mesh yet). If a tentative or intermediate mutation is required, then a panic should occur. In contrast, errors emitted when committing a mutation are non-recoverable, as that always means that the mesh has been mutated and is inconsistent. Maybe this could also be encoded into error types. It may also be possible to make them recoverable by cloning meshes before mutation (expensive).

**Comment by olson-sean-k** [comment]
This has proven to be a difficult approach to design. It seems the separation between a consistent user-facing API and an inconsistent internal API with non-trivial "guard rails" will introduce significant complexity. In any case, let's keep going. :-)

So far, encoding consistency into the `Mesh` and view types allows for different APIs to be expressed over the same type constructors. A somewhat different approach would be to decompose a `Mesh` for mutations, so that instead of going from `Mesh<..., Consistent>` to `Mesh<..., Inconsistent>`, we instead emit the underlying storage. That would allow for more flexibility over the (temporary) ownership of each type of topology (vertices, edges, and faces).

There are some benefits to doing this (not sure if they're worth the complexity yet), but one interesting requirement is the ability to build views on individual `Storage` references instead of entire `Mesh`es. [This gist](https://gist.github.com/olson-sean-k/4b7f6b6b50b30b19ec99d1bea3618f43) demonstrates a way to allow for this. A view can be built over a `Mesh`, enabling all of its features, but it can also be built over a reference to `Storage`. Additional references can be "fused", enabling more features, even if just temporarily over a local borrow. This works well if storage is not always grouped together and owned by a `Mesh`.

**Comment by olson-sean-k** [comment]
I've refined the approach described above: a type called `Core` now acts as a sort of ephemeral and polymorphic container for storage. An interesting consequence of being generic over storage for specific topologies (as opposed to an entire `Mesh`) is that view functions that operate on storage always take `self` and the semantics of this depend on the `AsStorage` input type (e.g., `Core`). For immutable views (e.g., `VertexRef`), these functions will just copy, but for mutable views (e.g., `VertexMut`) they will move.

Another interesting effect is that accessors can no longer use `mut` variants, including circulators. Because all functions take `self`, the differentiator is instead whether or not the result is a view or an orphan. For example, the pairing `incoming_edges + incoming_edges_mut` becomes `into_incoming_edges + into_incoming_orphan_edges`. Of course, mutations continue to require mutable storage access and consume `self`.

**Comment by olson-sean-k** [comment]
This has evolved a bit more. A good illustration is [this isolated code](https://github.com/olson-sean-k/plexus/blob/storage/src/graph/view/test.rs) for an alternative `VertexView` on the `storage` branch.

This begins to merge two important concepts: disjoint + ephemeral storage and encoding consistency into views. Disjoint storage brings great flexibility, including the ability to move ownership of storage into individual mutation types when mutating a mesh. It also allows for dynamically binding storage into a view as needed, using a borrow or otherwise. A consistency type parameter, applied only to views, means that internal code only has access to weaker APIs (the "reachable" API described earlier) when a mesh is being mutated and may be inconsistent.

All of this is invisible to user code. The `Ref`, `Mut`, and `Orphan` type aliases greatly simplify the exposed view types and user code will not interact with the internal mutation APIs. This should provide some useful structure for ensuring that internal code is correct and avoids exposing inconsistent topology to users.

Getting closer. :-)

**Comment by olson-sean-k** [comment]
The `storage` branch combines abstracted storage and mutations. That branch has reached a point where it now builds, passes tests, and runs examples. :tada:

The changes there add a considerable amount of complexity and I still need to consider carefully if this is the right approach. Here is a summary of what this enables:

- Storage for individual topologies is abstract (see the `AsStorage` trait). Views can be constructed over various types that provide storage, not just `Mesh`es. Chiefly, this allows storage for each topology to be moved out of a `Mesh` and owned by a mutation type that can guard that storage and provide limited read-only access to other types. When a mutation is committed, that storage is rebound into a `Mesh`.
- Consistency is encoded into view types, with only views constructed over `Mesh`es exposing operations that require consistency. Views expose a "reachable" API for any consistency, which is essentially "best effort" and fails gracefully. This API may also implement less efficient lookups in certain situations.
- The internal API is far more complex, but abstract storage facilitates a mutation API that allows for deferred checks for consistency and batch operations, like inserting multiple faces and detecting singularities when mutations are committed. User code is largely unaffected by this, and still relies on the view API, but internal code can consolidate mutations and error management within the mutation API. Panics within this API are always internal errors (bugs).

There are some open questions and tradeoffs as well. Here are some of the salient ones:

- The borrow checker more often complains about temporary views and lifetimes. It's not entirely clear to me why exactly this is, but it seems that the borrow checker doesn't reason about type parameters that may or may not be references the same way as concrete reference types. This worsens the ergonomics of some patterns, requiring awkward `map`ping or excessive `let` bindings. Interestingly, NLL seems to help immensly with this.
- These changes are complex, especially regarding views. All views now have three type parameters, and the parameter `M` used for storage is especially complex. There are many different constraints on these types used in many different `impl` blocks. This is definitely not trivial to understand and begin contributing to.
- Views are constructed in a way that relies on calling code to set the correct consistency. There are traits that can be used to ensure that a certain consistency is supported, but the bounds are complex, leak information about a function's implementation, and sometimes present impossible constraints due to cascading references with unnamed lifetimes. This is still an unsolved problem. (Maybe macros...?)

Regarding the borrow checker, some patterns frustratingly break down. For example, `unwrap` calls often lead to lifetime errors now.
```
let _ = mesh
    .vertex(key)
    .unwrap() // ERROR
    .incoming_edges()
    .map(|edge| edge.key())
    .collect::<Vec<_>>();
```
One way around this is to `map` and then `unwrap`, but this drifts rightwards.
```
let _ = mesh
    .vertex(key)
    .map(|vertex| {
        vertex
            .incoming_edges()
            .map(|edge| edge.key())
            .collect::<Vec<_>>()
    })
    .unwrap();
```
NLL appears to allow the original code.

**Comment by olson-sean-k** [comment]
It just occurred to me that the lifetime ergonomics problem is primarily caused by changes to circulators. Other call chains work as expected. For example:

```
let _ = mesh
    .vertex(key)
    .unwrap()
    .outgoing_edge()
    .key();
```

Circulators were also drastically changed while developing the `storage` branch. It may be possible to fix this problem...!

**Comment by olson-sean-k** [comment]
It seems that the use of `fn f() -> impl Iterator<...>` is [the problem](https://github.com/rust-lang/rust/issues/50823). Unfortunately, the new approach to circulators includes the use of a mapping iterator type with a type parameter that accepts a closure's unnameable type. The only way to fix this in the meantime would be to refactor the closures into local functions or revert to the original approach and explicitly name the iterator types in functions that emit them.

**Comment by olson-sean-k** [comment]
Almost ready to land this. The ergonomic issues concerning lifetimes have been fixed. The last change I'd like to make should help simplify view types while also eliminating problems with `from_keyed_storage` and transitive consistency.

Rather than establish consistency directly in view types, types that provide storage can implement a `Container` trait with an associated consistency type. The `C` parameter can be removed after adding a `M: Container` constraint. This further limits the types that can be accepted by views, bit the `Core` type is already as flexible and will implement `Container`. This should ultimately remove some "type algebra soup" and consistently apply consistency to views.

**Comment by olson-sean-k** [comment]
Commit cd0a7ee introduces the aforementioned `Container` trait as well as a new `Reborrow` trait. The `Container` trait works as described, and moves type-level data about consistency into the small number of types that provide aggregated mesh storage. The type parameter `C` in view types is now gone! :tada:

The `Reborrow` trait provides a generic way to avoid compounding references when generating views from other views. This unfortunately introduces it's own type complexity in the form of constraints on storage, but it means that irrespective of the storage type, references no longer compound. Ever. This is especially important for user-facing view type aliases (e.g., `FaceRef` or `VertexMut`), because they are based on `&Mesh<G>` as the storage type. If references compounded, that type may become `&&Mesh<G>`, and user code is unable to name such a view!

**Comment by olson-sean-k** [comment]
This is looking like a workable implementation, so I've landed the `storage` branch in `master`. :tada:

It still needs refinement, but that can be done on `master`. Landing this also unblocks other changes that I've been putting off.
